File: frontend/src/api/lab.ts
import { get, post, put, del } from './client'

export interface LabRequest {
  id: number
  tenant_id: string
  request_no: string
  title: string
  request_type: string
  source_type: string | null
  source_id: number | null
  priority: string
  sample_info: any
  description: string | null
  status: string
  assignee_id: number | null
  expected_date: string | null
  conclusion: string | null
  attachments: any
  created_by: number | null
  created_at: string
  updated_at: string
  test_results: LabTestResult[]
}

export interface LabTestResult {
  id: number
  request_id: number
  item_name: string
  spec_value: string | null
  actual_value: string | null
  unit: string | null
  lower_limit: number | null
  upper_limit: number | null
  is_pass: boolean | null
  remark: string | null
}

export interface TestStandard {
  id: number
  name: string
  category: string | null
  method: string | null
  default_lower_limit: number | null
  default_upper_limit: number | null
  unit: string | null
  description: string | null
  created_at: string
}

export interface LabCalibration {
  id: number
  equipment_name: string
  calibrate_type: string | null
  calibrate_date: string | null
  valid_until: string | null
  result: string | null
  certificate: string | null
  remark: string | null
}

export interface LabReport {
  id: number
  request_id: number
  report_no: string
  conclusion: string | null
  summary: string | null
  attachments: any
  published_by: number | null
  published_at: string | null
}

// 实验委托
export async function listRequests(params: {
  page?: number
  page_size?: number
  status?: string
  request_type?: string
  priority?: string
}): Promise<{ items: LabRequest[]; total: number; page: number; page_size: number }> {
  return get('/lab/requests', params)
}

export async function createRequest(data: any): Promise<LabRequest> {
  return post('/lab/requests', data)
}

export async function getRequest(id: number): Promise<LabRequest> {
  return get(`/lab/requests/${id}`)
}

export async function updateRequest(id: number, data: any): Promise<LabRequest> {
  return put(`/lab/requests/${id}`, data)
}

export async function receiveSample(id: number): Promise<LabRequest> {
  return post(`/lab/requests/${id}/receive`)
}

export async function assignTester(id: number, assignee_id: number): Promise<LabRequest> {
  return post(`/lab/requests/${id}/assign`, { assignee_id })
}

export async function startTesting(id: number): Promise<LabRequest> {
  return post(`/lab/requests/${id}/start`)
}

export async function submitResults(id: number, results: any[]): Promise<LabRequest> {
  return post(`/lab/requests/${id}/results`, { results })
}

export async function approveRequest(id: number): Promise<LabRequest> {
  return post(`/lab/requests/${id}/approve`)
}

export async function revertStatus(id: number): Promise<LabRequest> {
  return post(`/lab/requests/${id}/revert`)
}

export async function getReport(id: number): Promise<LabReport> {
  return get(`/lab/requests/${id}/report`)
}

export async function publishReport(id: number, conclusion: string, summary?: string, attachments?: string): Promise<LabReport> {
  return post(`/lab/requests/${id}/publish-report`, { conclusion, summary, attachments })
}

// 标准库
export async function listStandards(params: {
  page?: number
  page_size?: number
  category?: string
}): Promise<{ items: TestStandard[]; total: number; page: number; page_size: number }> {
  return get('/lab/standards', params)
}

export async function createStandard(data: any): Promise<TestStandard> {
  return post('/lab/standards', data)
}

export async function updateStandard(id: number, data: any): Promise<TestStandard> {
  return put(`/lab/standards/${id}`, data)
}

export async function deleteStandard(id: number): Promise<void> {
  return del(`/lab/standards/${id}`)
}

// 校准记录
export async function listCalibrations(params: {
  page?: number
  page_size?: number
}): Promise<{ items: LabCalibration[]; total: number; page: number; page_size: number }> {
  return get('/lab/calibrations', params)
}

export async function createCalibration(data: any): Promise<LabCalibration> {
  return post('/lab/calibrations', data)
}

File: backend/app/schemas/lab.py
# M15 实验室管理 — Pydantic Schemas
from pydantic import BaseModel, ConfigDict
from typing import Optional, Any, List
from datetime import datetime


# ── 实验委托 ────────────────────────────────────────────────

class LabRequestResponse(BaseModel):
    id: int
    tenant_id: str
    request_no: str
    title: str
    request_type: str
    source_type: Optional[str] = None
    source_id: Optional[int] = None
    priority: str = "medium"
    sample_info: Optional[Any] = None
    description: Optional[str] = None
    status: str = "pending"
    assignee_id: Optional[int] = None
    expected_date: Optional[datetime] = None
    conclusion: Optional[str] = None
    attachments: Optional[Any] = None
    created_by: Optional[int] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    test_results: List["LabTestResultResponse"] = []

    model_config = ConfigDict(from_attributes=True)


class CreateLabRequest(BaseModel):
    title: str
    request_type: str
    source_type: Optional[str] = None
    source_id: Optional[int] = None
    priority: str = "medium"
    sample_info: Optional[Any] = None
    description: Optional[str] = None
    expected_date: Optional[datetime] = None
    attachments: Optional[Any] = None


class UpdateLabRequest(BaseModel):
    title: Optional[str] = None
    request_type: Optional[str] = None
    source_type: Optional[str] = None
    source_id: Optional[int] = None
    priority: Optional[str] = None
    sample_info: Optional[Any] = None
    description: Optional[str] = None
    expected_date: Optional[datetime] = None
    conclusion: Optional[str] = None
    attachments: Optional[Any] = None


# ── 检测结果 ────────────────────────────────────────────────

class LabTestResultResponse(BaseModel):
    id: int
    tenant_id: str
    request_id: int
    item_name: str
    spec_value: Optional[str] = None
    actual_value: Optional[str] = None
    unit: Optional[str] = None
    lower_limit: Optional[float] = None
    upper_limit: Optional[float] = None
    is_pass: Optional[bool] = None
    remark: Optional[str] = None
    created_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class CreateTestResult(BaseModel):
    item_name: str
    spec_value: Optional[str] = None
    actual_value: Optional[str] = None
    unit: Optional[str] = None
    lower_limit: Optional[float] = None
    upper_limit: Optional[float] = None
    is_pass: Optional[bool] = None
    remark: Optional[str] = None


class SubmitResultsRequest(BaseModel):
    results: List[CreateTestResult]


class AssignRequest(BaseModel):
    assignee_id: int


# ── 检验标准库 ──────────────────────────────────────────────

class TestStandardResponse(BaseModel):
    id: int
    tenant_id: str
    name: str
    category: Optional[str] = None
    method: Optional[str] = None
    default_lower_limit: Optional[float] = None
    default_upper_limit: Optional[float] = None
    unit: Optional[str] = None
    description: Optional[str] = None
    created_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class CreateTestStandard(BaseModel):
    name: str
    category: Optional[str] = None
    method: Optional[str] = None
    default_lower_limit: Optional[float] = None
    default_upper_limit: Optional[float] = None
    unit: Optional[str] = None
    description: Optional[str] = None


class UpdateTestStandard(BaseModel):
    name: Optional[str] = None
    category: Optional[str] = None
    method: Optional[str] = None
    default_lower_limit: Optional[float] = None
    default_upper_limit: Optional[float] = None
    unit: Optional[str] = None
    description: Optional[str] = None


# ── 实验报告 ────────────────────────────────────────────────

class LabReportResponse(BaseModel):
    id: int
    tenant_id: str
    request_id: int
    report_no: str
    conclusion: Optional[str] = None
    summary: Optional[str] = None
    attachments: Optional[Any] = None
    published_by: Optional[int] = None
    published_at: Optional[datetime] = None
    created_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


# ── 校准记录 ────────────────────────────────────────────────

class LabCalibrationResponse(BaseModel):
    id: int
    tenant_id: str
    equipment_name: str
    calibrate_type: Optional[str] = None
    calibrate_date: Optional[datetime] = None
    valid_until: Optional[datetime] = None
    result: Optional[str] = None
    certificate: Optional[str] = None
    remark: Optional[str] = None
    created_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class CreateLabCalibration(BaseModel):
    equipment_name: str
    calibrate_type: Optional[str] = None
    calibrate_date: Optional[datetime] = None
    valid_until: Optional[datetime] = None
    result: Optional[str] = None
    certificate: Optional[str] = None
    remark: Optional[str] = None
